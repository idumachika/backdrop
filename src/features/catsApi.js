import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';

export const catsApi = createApi({
  reducerPath: 'catsApi',
  baseQuery: fetchBaseQuery({baseUrl: 'https://api.thecatapi.com/v1/'}),
  endpoints: builder => ({
    getAllCats: builder.query({
      query: () => 'breeds?limit=20',
    }),
  }),
});

export const {useGetAllCatsQuery} = catsApi;
