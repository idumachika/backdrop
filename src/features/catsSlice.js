import {createSlice} from '@reduxjs/toolkit';

export const catSlice = createSlice({
  name: 'cat',
  initialState: {
    favourite: false,
    cats: [],
  },
  reducers: {
    add(state, action) {
      const catExists = state.cats.find(item => item.id === action.payload.id);
      if (catExists) {
        return;
      }

      state.cats.push(action.payload);
    },
    remove(state, action) {
      state.cats = state.cats.filter(item => item.id !== action.payload);
    },
  },
});

export const {add, remove} = catSlice.actions;
export default catSlice.reducer;
