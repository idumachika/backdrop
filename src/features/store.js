import AsyncStorage from '@react-native-async-storage/async-storage';
import {configureStore} from '@reduxjs/toolkit';
import {persistReducer, persistStore} from 'redux-persist';
import catReducer from '../features/catsSlice';
import {catsApi} from './catsApi';

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
};

export const store = configureStore({
  reducer: {
    [catsApi.reducerPath]: catsApi.reducer,
    cats: persistReducer(persistConfig, catReducer),
  },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({serializableCheck: false}).concat(catsApi.middleware),
});

export const persistor = persistStore(store);
