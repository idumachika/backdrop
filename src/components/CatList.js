import * as React from 'react';
import {
  ActivityIndicator,
  FlatList,
  Image,
  Pressable,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import {heightPercentageToDP as hp} from 'react-native-responsive-screen';
import {useDispatch} from 'react-redux';
import Heart from '../assets/icons/Heart';
import {useGetAllCatsQuery} from '../features/catsApi';
import {add, remove} from '../features/catsSlice';
import {fonts} from '../theme/theme';

const CatList = () => {
  const {data, isError, error} = useGetAllCatsQuery();

  console.log('data-catlist', data);
  const renderItem = ({item}) => <Cat item={item} />;
  return (
    <>
      {isError ? (
        <Text>Error baby!!: {error?.message}</Text>
      ) : (
        <FlatList
          showsVerticalScrollIndicator={false}
          data={data}
          keyExtractor={item => item.id}
          renderItem={renderItem}
          ListEmptyComponent={() => (
            <ActivityIndicator size="large" color="#212227" />
          )}
          initialNumToRender={15}
        />
      )}
    </>
  );
};

const Cat = ({item}) => {
  const [favourite, setFavourite] = React.useState(false);
  const dispatch = useDispatch();
  // on load, show  which one is liked already
  const addCat = id => {
    if (id === item.id) {
      dispatch(add(item));
      setFavourite(true);
    }
  };

  const removeCat = id => {
    if (id === item.id) {
      dispatch(remove(id));
      setFavourite(false);
    }
  };

  return (
    <View style={styles.catContainer}>
      <View style={styles.leftSide}>
        <Image source={{uri: item?.image?.url}} style={styles.catImage} />
        <Text style={styles.catName}>{item?.name}</Text>
      </View>

      {favourite ? (
        <Pressable style={styles.likeBtn} onPress={() => removeCat(item.id)}>
          <Heart fill="#de0202" stroke="#de0202" />
        </Pressable>
      ) : (
        <Pressable style={styles.likeBtn} onPress={() => addCat(item.id)}>
          <Heart fill="none" stroke="#212227" />
        </Pressable>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  catContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: 25,
  },
  leftSide: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  catImage: {
    width: 60,
    height: 60,
    borderRadius: 10,
  },
  catName: {
    fontSize: hp('2.2'),
    lineHeight: 24,
    fontFamily: fonts.regular,
    paddingLeft: 15,
    color: '#212227',
  },
  likeBtn: {
    padding: 10,
  },
});
export default CatList;
