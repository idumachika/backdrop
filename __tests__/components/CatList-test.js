import React from 'react';
import CatList from '../../src/components/CatList';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react-native';
import { PersistGate } from 'redux-persist/integration/react';
import { persistor, store } from '../../src/features/store';



describe('<CatList/> tests', () => {
    test('should render the CatList', () => {
        const { toJSON } =
            render(<Provider store={store}>
                <PersistGate persistor={persistor}>
                    <CatList />
                </PersistGate>
            </Provider>);
        expect(toJSON()).toMatchSnapshot();


    });

    it('should fire onPress when pressed', () => {
        const onPressMock = jest.fn();
        const { queryByTestId } = render(
            <Provider store={store} >
                <PersistGate persistor={persistor}>
                    <CatList onPress={onPressMock} />
                </PersistGate>
            </Provider>
        );
        fireEvent.press(queryByTestId('CatList'));

        expect(onPressMock).toHaveBeenCalledTimes(1);
    });


});
