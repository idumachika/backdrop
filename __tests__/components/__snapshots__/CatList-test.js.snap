// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatList/> tests should render the CatList 1`] = `null`;
